src/dune_engine/targets.ml

```diff
@@ -190,6 +190,18 @@ module Produced = struct
       | Short_circuit -> None
   end
 
+  module Fiber = struct
+    let sequential_iteri t ~f =
+      let open Fiber.O in
+      let acc =
+        Path.Build.Map.foldi t.files ~init:(Fiber.return ())
+          ~f:(fun file payload acc -> acc >>> f file payload)
+      in
+      Path.Build.Map.foldi t.dirs ~init:acc ~f:(fun dir filenames acc ->
+          String.Map.foldi filenames ~init:acc ~f:(fun filename payload acc ->
+              acc >>> f (Path.Build.relative dir filename) payload))
```

Comment:
This style is not super efficient: by accumulating a `Fiber.t`, we basically start by accumulating a long chain of closures that we then execute at the end. Constructing an intermediate list might actually use less memory.

`Map` actually has a function that could help here: `Map.to_seq : 'a t -> (key * 'a) Seq.t`. We don't expose it in stdune at the moment, but it would be easy to add. This would allow a more streamed iteration where we construct fibers and execute them immediately.

Reply:
Hmm, interesting. 

Shall I use `Path.Build.Map.to_list (Targets.Produce.all_files t)` followed by `Fiber.sequential_iter`?

That's what I wrote initially but I wanted to avoid constructing an intermediate list. I always forget that closures aren't free :)

Reply:
I pushed a commit that does this.

Reply:
> Shall I use Path.Build.Map.to_list (Targets.Produce.all_files t) followed by Fiber.sequential_iter?

Using an intermediate list would probably create a more compact intermediate representation, but the Seq based version should be best as it will avoid creating a big intermediate data-structure altogether. Instead, we'll only allocated short lived values, which are much less likely to get promoted.

Reply:
Right, I see. I'll use a list for now but will add a CR-someday to expose `Map.to_seq` and switch to it in a separate PR.

Reply:
> Right, I see. I'll use a list for now but will add a CR-someday to expose Map.to_seq and switch to it in a separate PR.

I already pushed the PR to use a `Seq.t` :)

Reply:
Pushed the commit I mean

Reply:
Oh, I see, I didn't notice :) That's awesome, thanks!